assets/system/game_modules/kami.mjs

```diff
@@ -30,34 +30,87 @@
  *  POSSIBILITY OF SUCH DAMAGE.
 **/
 
-import from from 'from';
-
 export default new
 class Kami
 {
-	constructor(title = Sphere.Game.name)
+	constructor()
 	{
 		this.enabled = SSj.now() > 0;
-		this.title = title;
 		this.records = [];
+		this.outputBordered = false;
+		this.placeholder = new Record();
 		if (this.enabled)
 			this.exitJob = Dispatch.onExit(() => this.finish());
 	}
 
 	finish()
 	{
-		SSj.log(`Profiling has completed for "${this.title}"`);
-		let sortedRecords = from(this.records)
-			.where(it => it.count > 0)
-			.descending(it => it.totalTime / it.count);
-		for (const record of sortedRecords) {
-			let averageTime = Math.round(record.totalTime / record.count / 1000).toLocaleString();
-			let count = record.count.toLocaleString();
-			let totalTime = Math.round(record.totalTime / 1000).toLocaleString();
-			SSj.log(`  ${count} occurrences "${record.description}" took ${totalTime} mcs (avg. ${averageTime} mcs)`);
-			record.target[record.methodName] = record.originalFunction;
-		}
+		if (!this.enabled)
+			return;
+		
 		this.exitJob.cancel();
+
+		let totalTime = 0;
+		let totalAverage = 0;
+		for (let i = 0; i < this.records.length; ++i) {
+			if (this.records[i].count === 0) {
+				this.records.splice(i, 1);
+				--i;
+			}
+			else {
+				const record = this.records[i];
+				record.averageTime = record.totalTime / record.count;
+				totalTime += record.totalTime;
+				totalAverage += record.averageTime;
+			}
+		}
+
+		this.records.sort(function(a, b) { return b.averageTime - a.averageTime; });
+
+		let consoleOutput = [
+			[ "Event" ],
+			[ "Calls" ],
+			[ "Time (\u03bcs)" ],
+			[ "% Total" ],
+			[ "Avg (\u03bcs)" ],
+			[ "% Avg" ] ];
+		for (const record of this.records) {
+			consoleOutput[0].push(record.description);
+			consoleOutput[1].push(record.count.toLocaleString());
+			consoleOutput[2].push(Math.round(record.totalTime / 1000).toLocaleString());
+			consoleOutput[3].push(Math.round(100 * record.totalTime / totalTime) + "%");
+			consoleOutput[4].push(Math.round(record.averageTime / 1000).toLocaleString());
+			consoleOutput[5].push(Math.round(100 * record.averageTime / totalAverage) + "%");
+			if (record.attached === true)
+				record.target[record.methodName] = record.originalFunction;
+		}
+		SSj.log(`Profiling results for "${Sphere.Game.name}" total time: ${Math.round(totalTime/1000).toLocaleString()} \u03bcs` + makeTable(consoleOutput, this.outputBordered));
+		this.records.splice(0, this.records.length);
+	}
+
+	start(description = "Unknown")
+	{
+		if (!this.enabled)
+			return this.placeholder;
+		
+		let record = this.records.find((value) => {
```

Comment:
`.find()` feels kind of heavyweight for this; should we be using an object instead of an array to track records?

Reply:
Adding properties to an object can trigger a bailout this method should be consistent at least.

I can write an inlineable find equivalent instead I guess note this is done at the start of .start() before getting the time stamp - so only an issue for nested record blocks.

Reply:
Yeah, I'm not worried about skewing the results so much as introducing slowdown in the instrumented code - even if the timestamps are completely accurate, slow code is more annoying to test if nothing else.